acceptance/helpers.rb

```diff
@@ -41,21 +164,35 @@ def install_puppet_agent_module_on(host)
                                      module_name: 'puppet_agent')
       end
 
-      # Install puppet-agent on all agent nodes and connect them to the master.
-      # This is intended to prepare SUTs for tests that upgrade puppet-agent
-      # from some initial version.
+      # Read as "host_to_informative_string".
+      #
+      # @param [Beaker::Host] host The host
+      # @return An informative string of the form <host_name> (<host_platform>)
+      def host_to_info_s(host)
+        "#{host} (#{host['platform']})"
+      end
+
+      # Install puppet-agent on all agent nodes and connects them to the master.
+      # This is intended to setup SUTs for tests that upgrade puppet-agent
+      # from some initial version. A key post-condition of this method is that
+      # all of the agents are ready to be upgraded with the puppet_agent module.
       #
       # @param [String] initial_package_version_or_collection Either a version
       #   of puppet-agent or the name of a puppet collection to install the agent from.
-      def prepare_upgrade_with(initial_package_version_or_collection)
+      # @yield Invokes any additional set-up that's required. 
+      def setup_agents_to_upgrade(initial_package_version_or_collection)
```

Comment:
The reason why `purge_agents`' code was moved here and why this method now takes a block is because that is the best solution I could come up with that would ensure a clean SUT between each test, including _failed_ tests. The latter was a big pain point for me while working on this PR because I'd have to (manually) clean out my agent every time my acceptance tests failed (re-running the `rake prepare` task also took a while).

Reply:
Here were some of the other solutions I thought of. The existing way
```
setup_agents_to_upgrade('pc1')
teardown { purge_agents }
```

wouldn't work if one needed to do additional preparation before an upgrade. For example, given something like
```
setup_agents_to_upgrade('pc1')
// additonal prep here
teardown { purge_agents }
```
, if anything in the `additional prep` part of the code threw an exception, then the `purge_agents` teardown would not be included in Beaker's clean-up, so you'd need to manually go and clean your SUTs.

Something like
```
begin
  setup_agents_to_upgrade('pc1')
  // additional prep here
ensure
  teardown { purge_agents }
end
```

is also bad because (1) you need to remember to do that for every agent upgrade test to ensure a clean SUT and (2), if something in `setup_agents_to_upgrade` throws an exception, `purge_agents` will be called on agents that aren't properly configured.

Reply:
The current solution of
```
setup_agents_to_upgrade('pc1') do
   // additional prep here
end
```

is also not great because you need to remember to declare your prep. in a block, but outside of that one thing, you can code your prep. anyway you like since any teardowns inside the block are applied _before_ the code in `purge_agents`.

Reply:
Is it possible to include the teardown inside the setup_agents_to_upgrade function? https://github.com/puppetlabs/beaker/blob/a01b5732a7d6cec146bd18d08795c64aafe1d2f1/docs/concepts/style_guide.md#teardowns indicates there's no reason we need to wait until after the function completes to specify the teardown.

Then maybe a check in purge_agents to identify if there's an agent installed and skip the teardown if there isn't one?

Reply:
Addressed this in HipChat. Reason we need the purge_agents teardown at the end of the function is so that something like
```
setup_agents_to_upgrade('pc1') do
  <manipulate some state with Puppet>
  teardown { <restore that state with puppet> }
end
```

would work. Otherwise, when the "restore state" teardown is reached, the agent will have already been uninstalled.